### ISSUE TITLE
Phi-4 Dimensions Cut Problem

### ISSUE THREAD

**Issue by chwundermsft** [microsoft/PhiCookBook#350]
Hi,
I´m facing a pretty weird problem while doing a phi-4-mm finetuning. I´m using the transformers Trainer() to do the finetuning. Unfortunately, after loading a dataset, i put them into the processor. All processor outputs are being created correctly (input_ids, input_image_emebds, image_attention_mask, ...) and have the correct shape. I have printed that so you can also verify this. But as soon as i put that into the collator, it seems like it cuts all first dimensions of the input tensor. After that I get an error that the model expects a 5d image tensor, but got 4d.

My dataset is just a test dataset (stanford cars) which has the features "image, label". 

I have also used this code: https://[huggingface.co/microsoft/Phi-4-multimodal-instruct/resolve/main/sample_finetune_vision.py](https://huggingface.co/microsoft/Phi-4-multimodal-instruct/resolve/main/sample_finetune_vision.py) and in here it actually works, but I can´t find my error which is pretty weird as it should be all the same.

Here are all tensor shapes during the different phases of finetuning (dataset creation, before trainer, collator). 
In my samples below, you can ignore the full collator code, because the error happens already before any of this code!

`
working code - print inside dataset:
<class 'PIL.JpegImagePlugin.JpegImageFile'>
inputs.input_ids dataset: torch.Size([1, 523])
inputs.input_image_embeds dataset: torch.Size([1, 2, 3, 448, 448])
inputs.image_attention_mask dataset: torch.Size([1, 2, 32, 32])
inputs.image_sizes dataset: torch.Size([1, 2])

not working code - print inside dataset:
<class 'PIL.JpegImagePlugin.JpegImageFile'>
inputs.input_ids dataset: torch.Size([1, 549])
inputs.input_image_embeds dataset: torch.Size([1, 2, 3, 448, 448])
inputs.image_attention_mask dataset: torch.Size([1, 2, 32, 32])
inputs.image_sizes dataset: torch.Size([1, 2])

working code - print before trainer:
before trainer input_ids: torch.Size([1, 526])
before trainer labels: torch.Size([1, 526])
before trainer input_image_embeds: torch.Size([1, 2, 3, 448, 448])
before trainer image_attention_mask: torch.Size([1, 2, 32, 32])
before trainer image_sizes: torch.Size([1, 2])

not working code - print before trainer: 
before trainer input_ids: torch.Size([1, 802])
before trainer labels: torch.Size([1, 802])
before trainer input_image_embeds: torch.Size([1, 3, 3, 448, 448])
before trainer image_attention_mask: torch.Size([1, 3, 32, 32])
before trainer image_sizes: torch.Size([1, 2])

working code - inside the collator:
batch length: 1
input ids: torch.Size([1, 512])
labels: torch.Size([1, 512])
input_image_embeds: torch.Size([1, 2, 3, 448, 448])
image_attention_mask: torch.Size([1, 2, 32, 32])
image_sizes: torch.Size([1, 2])

not working code - inside the collator:
batch length: 1
input ids: torch.Size([561])
labels: torch.Size([561])
input_image_embeds: torch.Size([2, 3, 448, 448])
image_attention_mask: torch.Size([2, 32, 32])
image_sizes: torch.Size([2])
`

And here is the full code of training:

`
model_path = "microsoft/Phi-4-multimodal-instruct"
dtype = torch.bfloat16

processor = AutoProcessor.from_pretrained(model_path, trust_remote_code=True)

model = AutoModelForCausalLM.from_pretrained(
    model_path, 
    torch_dtype=dtype, 
    trust_remote_code=True,
    # if you do not use Ampere or later GPUs, change attention to "eager"
    _attn_implementation='flash_attention_2',
).to("cuda")

del model.model.embed_tokens_extend.audio_embed  # remove audio encoder
for layer in model.model.layers:
    # remove audio lora
    del layer.mlp.down_proj.lora_A.speech
    del layer.mlp.down_proj.lora_B.speech
    del layer.mlp.gate_up_proj.lora_A.speech
    del layer.mlp.gate_up_proj.lora_B.speech
    del layer.self_attn.o_proj.lora_A.speech
    del layer.self_attn.o_proj.lora_B.speech
    del layer.self_attn.qkv_proj.lora_A.speech
    del layer.self_attn.qkv_proj.lora_B.speech

model.set_lora_adapter('vision')
for param in model.model.embed_tokens_extend.image_embed.parameters():
    param.requires_grad = True

generation_config = GenerationConfig.from_pretrained(model_path)

user_prompt = '<|user|>'
assistant_prompt = '<|assistant|>'
prompt_suffix = '<|end|>'

class CarVisionDataset(Dataset):
    def __init__(self, processor):
        self.dataset = load_from_disk(os.getcwd()+"/dataset_train")
        self.processor = processor

    def __len__(self):
        return len(self.dataset)

    def __getitem__(self, idx):
        data = self.dataset[idx]
        image = data.get("image", None)
        label = data.get("label_name", None)
        
        prompt = [
            {"role": "system", "content": "Try to describe this car image with make model type and year?. Here is an exmaple: Bentley Continental Supersports Conv. Convertible 2012, BMW X6 SUV 2012, Chevrolet Corvette ZR1 2012, Chevrolet Silverado 1500 Classic Extended Cab 2007. Do not respond anything else!"},
            {"role": "user", "content": "<|image_1|>"}
        ]

        chat_template = self.processor.tokenizer.apply_chat_template(
            prompt,
            add_generation_prompt=True,
            tokenize=False
        )

        inputs = self.processor(
            text=chat_template,
            images=image,
            return_tensors="pt"
        )

        answer = f"{label}<|end|>\n<|endoftext|>"
        
        answer_ids = self.processor.tokenizer(answer, return_tensors='pt').input_ids
        
        input_ids = torch.cat([inputs.input_ids, answer_ids], dim=1)
        labels = torch.full_like(input_ids, -100)
        labels[:, -answer_ids.shape[1] :] = answer_ids

        if input_ids.size(1) > 8192:
            input_ids = input_ids[:, :8192]
            labels = labels[:, :8192]
            if torch.all(labels == -100).item():
                # workaround to make sure loss compute won't fail
                labels[:, -1] = self.processor.tokenizer.eos_token_id

        print(type(image))
        print(f"car dataset inputs.input_ids: {inputs.input_ids.shape}")
        print(f"cardataset labels: {labels.shape}")
        print(f"car dataset inputs.input_image_embeds: {inputs.input_image_embeds.shape}")
        print(f"car dataset inputs.image_attention_mask: {inputs.image_attention_mask.shape}")
        print(f"car dataset inputs.image_sizes: {inputs.image_sizes.shape}")
        
        return {
            'input_ids': input_ids,
            'labels': labels,
            'input_image_embeds': inputs.input_image_embeds,
            'image_attention_mask': inputs.image_attention_mask,
            'image_sizes': inputs.image_sizes,
        }

def custom_data_collator(batch):   

    print(len(batch))
    print(f"collator input_ids: {batch[0]['input_ids'].shape}")
    print(f"collator labels: {batch[0]['labels'].shape}")
    print(f"collator input_image_embeds: {batch[0]['input_image_embeds'].shape}")
    print(f"collator image_attention_mask: {batch[0]['image_attention_mask'].shape}")
    print(f"collator image_sizes: {batch[0]['image_sizes'].shape}")

    input_ids_list = []
    labels_list = []
    input_image_embeds_list = []
    image_attention_mask_list = []
    image_sizes_list = []
    for inputs in batch:
        input_ids_list.append(inputs['input_ids'])
        labels_list.append(inputs['labels'])
        input_image_embeds_list.append(inputs['input_image_embeds'])
        image_attention_mask_list.append(inputs['image_attention_mask'])
        image_sizes_list.append(inputs['image_sizes'])

    input_ids = pad_sequence(input_ids_list, padding_side='right', padding_value=0)
    labels = pad_sequence(labels_list, padding_side='right', padding_value=0)
    attention_mask = (input_ids != 0).long()
    input_image_embeds = cat_with_pad(input_image_embeds_list, dim=0)
    image_attention_mask = cat_with_pad(image_attention_mask_list, dim=0)
    image_sizes = torch.cat(image_sizes_list)
    
    return BatchFeature(
        {
            'input_ids': input_ids,
            'labels': labels,
            'attention_mask': attention_mask,
            'input_image_embeds': input_image_embeds,
            'image_attention_mask': image_attention_mask,
            'image_sizes': image_sizes,
            'input_mode': 1,  # vision mode
        }
    )

training_args = TrainingArguments(
    num_train_epochs=1,
    per_device_train_batch_size=1,
    gradient_checkpointing=True,
    gradient_checkpointing_kwargs={'use_reentrant': False},
    #gradient_accumulation_steps=gradient_accumulation_steps,
    optim='adamw_torch',
    adam_beta1=0.9,
    adam_beta2=0.95,
    adam_epsilon=1e-7,
    learning_rate=4.0e-5,
    weight_decay=0.01,
    max_grad_norm=1.0,
    lr_scheduler_type='linear',
    warmup_steps=50,
    logging_steps=10,
    output_dir='output',
    save_strategy='no',
    save_total_limit=10,
    save_only_model=True,
    fp16 = False,
    bf16 = True,
    remove_unused_columns=False,
    report_to='none',
    deepspeed=None,
    disable_tqdm=False,
    dataloader_num_workers=4,
    ddp_find_unused_parameters=True  # for unused SigLIP layers
)

from transformers import TrainerCallback
from transformers.integrations import MLflowCallback

train_vision = CarVisionDataset(processor)

print(f"before trainer input_ids: {train_vision[0]['input_ids'].shape}")
print(f"before trainer labels: {train_vision[0]['labels'].shape}")
print(f"before trainer input_image_embeds: {train_vision[0]['input_image_embeds'].shape}")
print(f"before trainer image_attention_mask: {train_vision[0]['image_attention_mask'].shape}")
print(f"before trainer image_sizes: {train_vision[0]['image_sizes'].shape}")

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_vision,
    data_collator=custom_data_collator
)

trainer.remove_callback(MLflowCallback)

trainer.train()

`

**Comment by leestott** [comment]
@chwundermsft 

Thanks for sharing the detailed breakdown of your issue with Phi-4-MM fine-tuning using Hugging Face's `Trainer`. Based on your description and the supporting data, the root cause appears to be a shape mismatch introduced during collation—specifically, the batch dimension is being dropped for image-related tensors like `input_image_embeds`, resulting in a 4D tensor instead of the expected 5D.

Here’s a synthesis of what I found across internal discussions, emails, and community threads that directly relate to your issue:

**Root Cause: Collator Misalignment with Tensor Shapes**

In your non-working case, the `input_image_embeds` tensor shape changes from `[1, 2, 3, 448, 448]` to `[2, 3, 448, 448]` inside the collator. This suggests that the batch dimension is being unintentionally squeezed or dropped during collation. The key insight is that `cat_with_pad` or similar logic might be flattening or misinterpreting the batch dimension when the batch size is 1.

**Working Example Reference**

The working version you referenced from Hugging Face’s official script (`sample_finetune_vision.py`) maintains the correct shape throughout. This script uses a custom collator that preserves the batch dimension explicitly and avoids any implicit tensor reshaping.

You can find a similar working pattern in the community discussion on Hugging Face at https://huggingface.co/microsoft/Phi-4-multimodal-instruct/discussions/ where users confirmed that using `unsqueeze(0)` or `torch.stack()` instead of `cat_with_pad()` helped maintain the correct shape.

**Suggested Fixes**

1. **Check `cat_with_pad()` Implementation**:
   Ensure that `cat_with_pad()` does not flatten the batch dimension. If you're using `torch.cat()` directly, wrap each tensor in a list and use `torch.stack()` instead:
   ```python
   input_image_embeds = torch.stack(input_image_embeds_list, dim=0)
   ```

2. **Guard Against Batch Size 1 Edge Case**:
   When batch size is 1, `torch.squeeze()` or similar operations can inadvertently drop dimensions. Add a check to preserve shape:
   ```python
   if input_image_embeds.dim() == 4:
       input_image_embeds = input_image_embeds.unsqueeze(0)
   ```

3. **Use `default_data_collator` as Baseline**:
   If you're customizing the collator, start from `transformers.default_data_collator` and extend it to handle multimodal inputs.
